File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def __init__(self):
        self.copies = {}

    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head: return None
        temp = head
        while head:
            self.copies[head] = Node(head.val)
            head = head.next
        head = temp
        while head:
            self.copies[head].next = self.copies[head.next] if head.next else None
            self.copies[head].random = self.copies[head.random] if head.random else None
            head = head.next
        return self.copies[temp]

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    if head == nil { return nil }
    curr := head
    for curr != nil {
        copy := Node{curr.Val,nil,nil}
        copy.Next = curr.Next
        curr.Next = &copy
        curr = curr.Next.Next
    }
    curr = head 
    for curr != nil {
        temp := curr.Next
        if curr.Random != nil {
            temp.Random = curr.Random.Next
        }
        curr = curr.Next.Next
    }
    newHead := head.Next
    curr = head 
    for curr != nil {
        temp := curr.Next
        curr.Next = temp.Next
        if temp.Next != nil {
            temp.Next = temp.Next.Next
        }
        curr = curr.Next
    }
    return newHead
}